plugin/src/main/java/io/jenkins/plugins/casc/core/HudsonPrivateSecurityRealmConfigurator.java

```diff
@@ -3,43 +3,78 @@
 import io.jenkins.plugins.casc.Attribute;
 import io.jenkins.plugins.casc.impl.attributes.MultivaluedAttribute;
 import io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator;
+import jenkins.model.Jenkins;
 import org.kohsuke.accmod.Restricted;
 import org.kohsuke.accmod.restrictions.NoExternalUse;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import java.io.IOException;
+import java.util.Collection;
 import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import java.util.stream.Collectors;
 
 /**
  * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
  */
 @Extension
 @Restricted(NoExternalUse.class)
-public class HudsonPrivateSecurityRealmConfigurator extends DataBoundConfigurator {
+public class HudsonPrivateSecurityRealmConfigurator extends DataBoundConfigurator<HudsonPrivateSecurityRealm> {
+    private static final Logger logger = Logger.getLogger(HudsonPrivateSecurityRealmConfigurator.class.getName());
+    private static final VersionNumber MIN_VERSION_FOR_HASHED_PASSWORDS = new VersionNumber("2.161");
+    /// matches HudsonPrivateSecurityRealm.JBCRYPT_HEADER
+    private static final String HASHED_PASSWORD_PREFIX = "#jbcrypt:";
 
     public HudsonPrivateSecurityRealmConfigurator() {
         super(HudsonPrivateSecurityRealm.class);
     }
 
     @NonNull
     @Override
-    public Set<Attribute> describe() {
-        final Set<Attribute> describe = super.describe();
+    public Set<Attribute<HudsonPrivateSecurityRealm, ?>> describe() {
+        final Set<Attribute<HudsonPrivateSecurityRealm, ?>> describe = super.describe();
         describe.add(new MultivaluedAttribute<HudsonPrivateSecurityRealm, UserWithPassword>("users", UserWithPassword.class)
-            .getter(target ->
-                target.getAllUsers().stream()
-                    .map(u -> new UserWithPassword(u.getId(), null))    // password isn't actually stored, only hashed
-                    .collect(Collectors.toList()))
-            .setter((target, value) -> {
-                for (UserWithPassword user : value) {
+            .getter(HudsonPrivateSecurityRealmConfigurator::getter)
+            .setter(HudsonPrivateSecurityRealmConfigurator::setter)
+        );
+        return describe;
+    }
+
+    private static Collection<UserWithPassword> getter(HudsonPrivateSecurityRealm target) {
+        return target.getAllUsers().stream()
+                .map(u -> new UserWithPassword(u.getId(), null)) // password isn't actually stored, only hashed
+                .collect(Collectors.toList());
+    }
+
+    private static void setter(HudsonPrivateSecurityRealm target, Collection<UserWithPassword> value) throws IOException {
+        for (UserWithPassword user : value) {
+            if (user.password.startsWith(HASHED_PASSWORD_PREFIX) && jenkinsSupportsHashedPasswords()) {
+                try {
+                    target.createAccountWithHashedPassword(user.id, user.password);
```

Comment:
we need to be able to compile with 2.60.3, yet signature is not there according to current target :sweat_smile:

Reply:
Ya...in my original Jenkins PR I had made it so that the same method signature was used, it just detected the jbcrypt header and did magic. That was denied.

I guess I can use reflection here, which feels weird unless you know of a better way.

Reply:
Reflection is the way to go